### ISSUE TITLE
Word bigram/trigram component separate frequency checking

### ISSUE THREAD

**Issue by martinreynaert** [LanguageMachines/ticcltools#4]
In TICCL-ldcalc I need an extra filter on pairs of ngrams. For bi- and trigrams we usually do not have validated forms, i.e. frequency counts equal or higher than the artificial frequency (artifrq). Therefore, the frequencies of their component unigrams need to be checked separately and if they are all equal to or higher than the artifrq, the pair needs to be blocked (filtered out).

E.g. 'a_pair_of' vs. 'a_pear_of': all six component words are validated words: remove
But: 'a_pair_of' vs. 'a_piar_of': retain

**Comment by kosloot** [comment]
Probably it is better to solve this in TICCL-unk.
just give 'clean' ngrams a frequency incremented by artifreq.

such a change is in test now.

**Comment by kosloot** [comment]
@martinreynaert please comment on this. does it work as expected?

**Comment by martinreynaert** [comment]
For now, let us assume that this indeed works as intended. As far as I have seen it does. Neverthess, all is not well.

Also, as far as I can see now, it is right that TICCL-unk should handle this: it allows superusers to prepare background files for use by common users.

In my current analysis, this fails in two subtle ways, depending on how the background file is used, i.e. either given to TICCL-unk or given to TICCL-anahash later.

When the background file is given to TICCL-unk, it can validate ngrams, as it should do. However, all the non-validated ngrams present only in the background file will also be incorporated by TICCL-anahash in the foci file and consequently handled by TICCL-indexer(NT) and TICCL-LDcalc, etc. This will result in far too much work being done and time consumed for the common user who just wants the one book corrected.

When the background file is given to TICCL-anahash, TICCL-unk does not get the necessary frequency data to validate many ngrams in the single book's frequency lists and so very many ngrams that would otherwise be discarded are also incorporated in the foci list by TICCL-anahash, resulting in very many False Positives later. TICCL-anahash does not repeat the frequency validation on the basis of the background corpus performed by TICCL-unk, although it does work on frequencies in order to build the foci list. Neither does TICCL-LDcalc repeat this, although I always assumed it did check otherwise capitalised word frequencies for unigrams, but this is then probably another issue we shall have to look into as far as the bi- and trigrams are concerned.

What we do need to keep in mind is that we have two types of users: superusers who need to process large corpora and are prepared to wait and common users who just want their small job done as quickly as possible. 

Also: if ever we want to be able to correct confusables, we need to be able to let the system examine the complete ngram lists of the input corpus to be corrected. Let us further call this the foreground corpus.

One possible way out that I see is to let TICCL-anahash only incorporate ngrams that actually are present in the foreground corpus in the foci list. When no foci list is built, TICCL-indexer(NT) should take the full foreground corpus frequency lists anagram values as foci file. Another related issue that I have raised before is that this list should be the only word forms TICCL-LDcalc examines and not also all the other anagrams connected to the anagram values in the foci list.

The above entails some design decisions we need to carefully consider now. However, in order to finally get anywhere with TICCL-ngram correction, I hope we can nail all this soon.

MRE

**Comment by kosloot** [comment]
I see one thing I suggest to do right now:
adapting TICC-anahash as suggested, to make the foci file shorter (maybe as an option?)

Concerning this remark:
``When no foci list is built, TICCL-indexer(NT) should take the full foreground corpus frequency lists anagram values as foci file.``

- this is only relevant for indexer.  indexerNT requires a foci-file
- as far as i understand it, indexer uses the foreground file already to work on. So what do you mean?

**Comment by martinreynaert** [comment]
Please adapt TICCL-anahash as suggested as quickly as possible. If you make it an option, please make it the default.

As regards the 'no' foci file remark: this was badly put. I should probably have written: 'I would like the option to have TICCL-indexer(NT) examine the full frequency listst of the foreground corpus without any of the background corpus only anagram hash values being incorporated in the foci file.'

**Comment by kosloot** [comment]
Ok, starting to look into this and getting confused directly.
your suggestion was:

``One possible way out that I see is to let TICCL-anahash only incorporate ngrams that actually are present in the foreground corpus in the foci list. ``

But isn't the foreground corpus exactly the input file for TICCL-anahash?
anahash wiil not add entries to the foci, that are NOT present in that inputfile.
Or did you mean checking against the **background** file?

For future bug reports it would be handy to have a SMALL test dataset, were you clearly point out:
''Giving _this_ inputfiles, and running _that_ program, i wouldn't expect _such_ outcome.''
For individual programs, the whole PICCL flow is way to convoluted for me to debug individual parts.

**Comment by kosloot** [comment]
some new questions arise here, to clarify my thoughts.
1. isn't the background file supposed to be a clean file per definition?
  shouldn't in fact, TICCL-unk produce a mere copy of this file as .clean and an empty .punct file when run on the background only?
2. so 'filtering' on the foreground corpus more or less comes down to taking a subset of the background, limiting ourself to the words in focus.
3. if this is so, then the simplest solution is to build 2 clean lists in parallel.
- when a foreground word is clean it is added to .clean
- when it ALSO appears in the background, it is added to an extra file background.clean
- there is no need for an extra .punct file. as the background never contributes to that.

**Comment by kosloot** [comment]
in fact filenames could be different/smarter as the backgound.clean file is used for further processing in anahash to get the foci.

**Comment by martinreynaert** [comment]
In the chat we just came to the conclusion that the problem might perhaps be solved either by running TICCL-unk twice (which would entail taking care of not overwriting prior output files) or by perhaps adding a parameter and minimal modifications to the module.

Let us examine prior output form TICCL-unk run on the English MORSE book (CLARIAH pilot project OpenGazam).

In RUNAUTO3 the background file was added through TICCL-anahash, not available to TICCL-unk. In RUNAUTO4 the background file was also available to TICCL-unk.

File sizes:

reynaert@red:/reddata/PILOTS/MORSE$ ls -l /reddata/PILOTS/MORSE/RUNAUTO3/zzz/TICCL/*clean
-rw-r--r-- 1 reynaert reynaert 6831824 Jul 13 01:02 /reddata/PILOTS/MORSE/RUNAUTO3/zzz/TICCL/RUNAUTO3.wordfreqlist.1to3.tsv.clean
reynaert@red:/reddata/PILOTS/MORSE$ ls -l /reddata/PILOTS/MORSE/RUNAUTO3/zzz/TICCL/*punct
-rw-r--r-- 1 reynaert reynaert 6794370 Jul 13 01:02 /reddata/PILOTS/MORSE/RUNAUTO3/zzz/TICCL/RUNAUTO3.wordfreqlist.1to3.tsv.punct
reynaert@red:/reddata/PILOTS/MORSE$ ls -l /reddata/PILOTS/MORSE/RUNAUTO3/zzz/TICCL/*anahash
-rw-r--r-- 1 reynaert reynaert 367580325 Jul 13 01:04 /reddata/PILOTS/MORSE/RUNAUTO3/zzz/TICCL/RUNAUTO3.wordfreqlist.1to3.tsv.clean.anahash

reynaert@red:/reddata/PILOTS/MORSE$ ls -l /reddata/PILOTS/MORSE/RUNAUTO4/zzz/TICCL/*clean
-rw-r--r-- 1 reynaert reynaert 302091715 Jul 14 14:47 /reddata/PILOTS/MORSE/RUNAUTO4/zzz/TICCL/RUNAUTO4.wordfreqlist.1to3.tsv.clean
reynaert@red:/reddata/PILOTS/MORSE$ ls -l /reddata/PILOTS/MORSE/RUNAUTO4/zzz/TICCL/*punct
-rw-r--r-- 1 reynaert reynaert 6794667 Jul 14 14:47 /reddata/PILOTS/MORSE/RUNAUTO4/zzz/TICCL/RUNAUTO4.wordfreqlist.1to3.tsv.punct
reynaert@red:/reddata/PILOTS/MORSE$ ls -l /reddata/PILOTS/MORSE/RUNAUTO4/zzz/TICCL/*anahash
-rw-r--r-- 1 reynaert reynaert 367581127 Jul 14 14:49 /reddata/PILOTS/MORSE/RUNAUTO4/zzz/TICCL/RUNAUTO4.wordfreqlist.1to3.tsv.clean.anahash

In RUNAUTO3 the clean file is very noticeably smaller. The other files differ a little. The background file had been run through TICCL-unk in a preparatory step, in order to make it 'clean' for subsequent processing runs involving English.

The clean output of RUNAUTO3 in fact contains all the items that given their frequencies are below artifrq, should by TICCL-anahash be put into the foci file.

Let is now examine the 'collected' frequencies in both clean-files. I search for a 'known' confusable for the bigram 'the bay' where the OCR has turned 'tire' (an existing word which in its meaning of 'wheel tire' is currently spelled as 'tyre' in British English, but this is an aside).

reynaert@red:/reddata/PILOTS/MORSE$ grep 'tire_bay' /reddata/PILOTS/MORSE/RUNAUTO3/zzz/TICCL/*clean
tire_bay	4
of_tire_bay	2
into_tire_bay	1
through_tire_bay	1
tire_bay_is	1
tire_bay_lies	1
tire_bay_of	1
tire_bay_towards	1
reynaert@red:/reddata/PILOTS/MORSE$ 
reynaert@red:/reddata/PILOTS/MORSE$ grep 'tire_bay' /reddata/PILOTS/MORSE/RUNAUTO4/zzz/TICCL/*clean
tire_bay	1000000004
of_tire_bay	1000000002
into_tire_bay	1000000001
through_tire_bay	1000000001
tire_bay_is	1000000001
tire_bay_lies	1000000001
tire_bay_of	1000000001
tire_bay_towards	1000000001
entire_bay	39

We see that given the background corpus, all ngrams containing 'tire_bay' have been internally validated by TICCL-unk. Which is as should be. They will not currently be written to the foci file. 

Due to lack of the background validated artifrqs, these ngrams are not validated in AUTORUN3. They are written to the foci file. Which is in a way correct, but we are not yet trying to solve confusables in TICCL. (Much as we woulld like to...).

What is definitely not correct is that the equally valid 'entire_bay' bigram is not validated in AUTORUN4. This is definitely a bug in the current version. We see it does not appear in the foreground corpus. Incorporated from the background corpus and not being validated, it will also (and many thousands of others like it) written to the foci file by TICCL-anahash. All these hugely contribute to the problem we are currently trying to address.

More asap.

MRE

**Comment by martinreynaert** [comment]
Regarding the 'entire_bay' case...

We see that the foci files differ greatly:

reynaert@red:/reddata/PILOTS/MORSE$ ls -l /reddata/PILOTS/MORSE/RUNAUTO3/zzz/TICCL/RUNAUTO3.wordfreqlist.1to3.tsv.clean.corpusfoci
-rw-r--r-- 1 reynaert reynaert 4560513 Jul 13 01:02 /reddata/PILOTS/MORSE/RUNAUTO3/zzz/TICCL/RUNAUTO3.wordfreqlist.1to3.tsv.clean.corpusfoci
reynaert@red:/reddata/PILOTS/MORSE$ ls -l /reddata/PILOTS/MORSE/RUNAUTO4/zzz/TICCL/RUNAUTO4.wordfreqlist.1to3.tsv.clean.corpusfoci
-rw-r--r-- 1 reynaert reynaert 24630797 Jul 14 14:49 /reddata/PILOTS/MORSE/RUNAUTO4/zzz/TICCL/RUNAUTO4.wordfreqlist.1to3.tsv.clean.corpusfoci

Less of the AUTORUN3 foreground corpus have been validated, which may account for the greater number of focus items there. 

However, we expected to see non-validated bigrams such as 'entire_bay' in the AUTORUN4 foci list. 

We do not see them (and in the meantime remember why...):

reynaert@red:/reddata/PILOTS/MORSE$ grep 'entire_bay' /reddata/PILOTS/MORSE/RUNAUTO4/zzz/TICCL/RUNAUTO4.wordfreqlist.1to3.tsv.clean.anahash
156444599706~by_trainee#entire_Bay#entire_bay#trainee_by

Check in foci lists:
reynaert@red:/reddata/PILOTS/MORSE$ grep '156444599706' /reddata/PILOTS/MORSE/RUNAUTO3/zzz/TICCL/RUNAUTO3.wordfreqlist.1to3.tsv.clean.corpusfoci
reynaert@red:/reddata/PILOTS/MORSE$ grep '156444599706' /reddata/PILOTS/MORSE/RUNAUTO4/zzz/TICCL/RUNAUTO4.wordfreqlist.1to3.tsv.clean.corpusfoci
reynaert@red:/reddata/PILOTS/MORSE$

Indeed: not there. This is due to the extra frequency check performed by TICCL-anahash. That seems OK, then.

I nevertheless need to look further: I seem to have seen such cases unexpectedly appear in TICCL-rank output. Maybe some such are nevertheless retrieved by TICCL-LDcalc, somehow...

**Comment by martinreynaert** [comment]
In conclusion from all my ramblings above:

What I think is required now:

- That TICCL-unk is able to produce a separate list containing only the ngrams present in the foreground corpus
- That this list is processable by TICCL-anahash, apart from the full clean file, to create a separate anahash file that is to serve as the foci list, based on the artifrq as before. This list will have only the ngrams available in the foreground corpus in contrast to the full anahash. Having the actual ngrams in the foci list as well, in contrast to before, will make it a lot easier to verify things.
- That TICCL-LDcalc is guided by this list in order to create the list of word variants: i.e. no word or word combination that is not in the foreground corpus should ever be proposed as a variant for Correction Candidates. This may well happen now as this is determined by the frequencies of the words in the pair variant vs. Correction Candidate. This frequency ordering mechanism should nevertheless stay in place as is.

MRE

**Comment by kosloot** [comment]
This discussion is (too) long, confusing and partly outdated.
I close this. For issues still relevant, I suggest opening separate new issues.
